apps/web/vibes/soul/form/file-upload/file-upload-dropzone.tsx

```diff
@@ -0,0 +1,189 @@
+'use client';
+
+import { clsx } from 'clsx';
+import { UploadIcon } from 'lucide-react';
+import { useCallback, useEffect } from 'react';
+
+import type { FileUploadProps } from '@/vibes/soul/form/file-upload';
+import { useFileUpload } from '@/vibes/soul/form/file-upload/file-upload-provider';
+import { Label } from '@/vibes/soul/form/label';
+import { Button } from '@/vibes/soul/primitives/button';
+
+export interface FileUploadDropzoneProps
+  extends Pick<
+    FileUploadProps,
+    | 'accept'
+    | 'register'
+    | 'registerProps'
+    | 'multiple'
+    | 'name'
+    | 'required'
+    | 'label'
+    | 'hideLabel'
+    | 'cta'
+    | 'message'
+  > {
+  label?: string;
+  message?: string;
+}
+
+export function FileUploadDropzone({
+  accept,
+  multiple,
+  name,
+  label,
+  cta = 'Upload file',
+  message = 'or drag and drop files here',
+  required,
+  hideLabel,
+  register,
+  registerProps,
+}: FileUploadDropzoneProps) {
+  const {
+    state,
+    inputRef,
+    hiddenInputRef,
+    onDragEnter,
+    onDragLeave,
+    onDragOver,
+    onDrop,
+    onPaste,
+    onKeyDown,
+    onDropzoneClick,
+    onInputChange,
+    disabled,
+    labelId,
+    inputId,
+    listId,
+    dropzoneId,
+    onTriggerClick,
+  } = useFileUpload();
+
+  const { dragOver, invalid } = state;
+
+  const setHiddenInputRef = useCallback(
+    (element: HTMLInputElement | null) => {
+      hiddenInputRef.current = element;
+      if (typeof register === 'function') {
+        register(element);
+      }
+    },
+    [register, hiddenInputRef],
+  );
+
+  useEffect(() => {
+    if (!hiddenInputRef.current) return;
+
+    const syncFiles = () => {
+      if (!hiddenInputRef.current) return;
+
+      const filesInStore = Array.from(state.files.values()).map((f) => f.file);
+
+      const dataTransfer = new DataTransfer();
+      for (const file of filesInStore) {
+        dataTransfer.items.add(file);
+      }
+
+      hiddenInputRef.current.files = dataTransfer.files;
+    };
+
+    syncFiles();
+  }, [state, hiddenInputRef]);
```

Comment:
What does this hook do? Is it synchronizing the state with the DOM (i.e., file input)? It might be helpful to extract this to a custom hook and give it some description with JSDoc so that it's clear what it's doing.

That said, is there a reason we didn't derive this state from the input directly instead of tracking it in React state?